torch/_dynamo/optimizations/training.py

```diff
@@ -260,42 +260,139 @@ def __call__(self, gm: torch.fx.GraphModule, example_inputs):
 aot_mem_efficient_fusion_no_decomp = AOTMemEfficientFusionWithContext(False)
 
 
-def prims_executor(gm, inputs, *, executor):
-    # This function is called once per forward/backward pass of a graph in AOT
-    # Autograd. We use it to set up the nvFuser-specific FX graph and return
-    # execute function.
+def _has_incompatible_cudagraph_ops(gm):
+    from torch._inductor.utils import (
+        has_incompatible_cudagraph_ops as has_incompatible_cudagraph_ops_inductor,
+    )
+
+    result = has_incompatible_cudagraph_ops_inductor(gm)
+    if result:
+        return result
+
+    incompatible = [
+        "aten.index_put.default",
+    ]
+    for node in gm.graph.nodes:
+        if str(node.target) in incompatible:
+            return True
+
+    # A workaround for
+    # https://github.com/csarofeen/pytorch/issues/2106
+    is_any_ones = any(
+        node.target == torch.ops.aten.ones.default for node in gm.graph.nodes
+    )
+    is_any_zeros = any(
+        node.target == torch.ops.aten.zeros.default for node in gm.graph.nodes
+    )
+    if is_any_ones and is_any_zeros:
+        return True
+
+    return False
+
+
+def _replace_cpu_with_cuda(gm):
+    for node in gm.graph.nodes:
+        if node.op == "call_function":
+            new_kwargs = dict(node.kwargs)
+            if new_kwargs.get("device", False) and new_kwargs["device"].type == "cpu":
+                new_kwargs["device"] = torch.device("cuda")
+            node.kwargs = new_kwargs
+    gm.recompile()
+    return gm
+
+
+def prims_executor(gm, inputs, *, executor, num_fixed=0, cudagraphs=False):
+    """This function is called once per forward/backward pass of a graph in AOT
+    Autograd. We use it to set up the nvFuser-specific FX graph and return
+    execute function."""
+    from functorch.compile import make_boxed_func
+
+    from torch._inductor.compile_fx import align_inputs, cudagraphify
     from torch._prims.context import TorchRefsNvfuserCapabilityMode
     from torch._prims.executor import execute
     from torch.fx.experimental.proxy_tensor import make_fx
 
+    # cudagraphify fails if intermediate tensors are CPU and there's an attempt
+    # to send them to CUDA
+    if cudagraphs:
+        gm = _replace_cpu_with_cuda(gm)
+
     # First we trace the graph conditionally decomposing nodes
     # that can be sent to the nvfuser executor
     with TorchRefsNvfuserCapabilityMode():
         prim_gm = make_fx(gm)(*inputs)
 
     # Then we return a callable that executes the "prim_gm" graph
-    return partial(execute, prim_gm, executor=executor)
+    run = make_boxed_func(partial(execute, prim_gm, executor=executor))
 
+    if _has_incompatible_cudagraph_ops(prim_gm) or not cudagraphs:
+        return run
 
-def create_nvprims_backend(*, executor):
+    try:
```

Comment:
nitpick: context manager looks cleaner than a try/finally.

Reply:
But context managers are implemented with try: finally:
I didn't do that with a context manager because the code is not reused. I can still change that of course.

Reply:
This is just a style nitpick :smiley: It's your call.